### PR TITLE
Eliminating use of acceptMsgDirectTo [6277]

### PR DESCRIPTION
**Pull request by araccp**
Optimization to reduce calls to EntityId_t::operator== by converting
MessageReceiver::AssociatedReaders from a vector<RTPSReader*> to an
unordered_map<EntityId_t, vector<RTPSReader*>>.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/181/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/166/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/180/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/203/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/185/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/202/)

**Comment by MiguelCompany** [comment]
I think this could be a nice improvement.

It seems the test failures are all related to the MessageReceiver destructor complaining that `AssociatedReaders` is not empty. 

We could remove the entry from the map when the vector gets emptied on `removeEndpoint`. This way we may also remove the check for empty on `willAReaderAcceptMsgDirectedTo`

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/223/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/202/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/222/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/264/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/240/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/263/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/306/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/281/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/304/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/309/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/284/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/307/)

**Comment by MiguelCompany** [comment]
I will investigate the regressions, as some tests are throwing SEGFAULT errors

File: src/cpp/rtps/messages/MessageReceiver.cpp [hunk 1]
```diff
@@ -406,6 +419,47 @@ bool MessageReceiver::readSubmessageHeader(CDRMessage_t* msg, SubmessageHeader_t
...
+
+    logWarning(RTPS_MSG_IN, IDSTRING"No Reader accepts this message (directed to: " <<readerID << ")");
+    return false;
+}
+
+void MessageReceiver::findAllReaders(const EntityId_t & readerID, std::function<void(RTPSReader*)> callback)
+{
+    if (readerID != c_EntityId_Unknown) {
+        const auto readers = AssociatedReaders.find(readerID);
+        for (const auto & it : readers->second) callback(it);
```
Comment (MiguelCompany) [inline src/cpp/rtps/messages/MessageReceiver.cpp]: `readers` should be checked to be different from AssociatedReaders.end() before accessing `->second`

File: src/cpp/rtps/messages/MessageReceiver.cpp [hunk 2]
```diff
@@ -406,6 +419,47 @@ bool MessageReceiver::readSubmessageHeader(CDRMessage_t* msg, SubmessageHeader_t
...
+bool MessageReceiver::willAReaderAcceptMsgDirectedTo(const EntityId_t & readerID)
+{
+    if(AssociatedReaders.empty()) {
+        logWarning(RTPS_MSG_IN,IDSTRING"Data received when NO readers are listening");
+        return false;
+    }
+
+    if (readerID != c_EntityId_Unknown) {
+        const auto readers = AssociatedReaders.find(readerID);
+        if (readers != AssociatedReaders.end() && !readers->second.empty()) {
```
Comment (MiguelCompany) [inline src/cpp/rtps/messages/MessageReceiver.cpp]:
```suggestion
        if (readers != AssociatedReaders.end()) {
```
As we are now removing the entry when the vector becomes empty, the last check is not necessary

**Comment by MiguelCompany** [comment]
Closing this in favor of #705

**Pull request closed without merging**